src/core/test_helpers/so_migrations.ts

```diff
@@ -0,0 +1,175 @@
+      );
+    }
+  };
+
+  /**
+   * Exports objects from Kibana with all migrations applied.
+   * @param options
+   */
+  const exportObjects = async (options: ExportOptions): Promise<SavedObject[]> => {
+    if (!started)
+      throw new Error(`SavedObjectTestHarness must be started before objects can be imported`);
+    if (stopped) throw new Error(`SavedObjectTestHarness cannot import objects after stopped`);
+
+    const response = await kbnTestServer
+      .getSupertest(root, 'post', '/api/saved_objects/_export')
+      .send({
+        ...options,
+        excludeExportDetails: true,
+      })
+      .expect(200);
+
+    // Parse ndjson response
+    return response.text.split('\n').map((s: string) => JSON.parse(s));
+  };
+
+  return {
+    /**
+     * Start Kibana and Elasticsearch for migration testing. Must be called before `migrate`.
+     * In most cases, this can be called during your test's `beforeAll` hook and does not need to be called for each
+     * individual test.
+     */
+    start: async () => {
+      if (started)
+        throw new Error(`SavedObjectTestHarness already started! Cannot call start again`);
+      if (stopped)
+        throw new Error(`SavedObjectTestHarness already stopped! Cannot call start again`);
+
+      started = true;
+      esServer = await startES();
+      await root.setup();
+      await root.start();
+
+      await waitForTrue(async () => {
+        const statusApi = kbnTestServer.getSupertest(root, 'get', '/api/status');
+        const response = await statusApi.send();
+        return response.status === 200;
+      });
+    },
+
+    /**
+     * Stop Kibana and Elasticsearch for migration testing. Must be called after `start`.
+     * In most cases, this can be called during your test's `afterAll` hook and does not need to be called for each
+     * individual test.
+     */
+    stop: async () => {
+      if (!started) throw new Error(`SavedObjectTestHarness not started! Cannot call stop`);
+      if (stopped)
+        throw new Error(`SavedObjectTestHarness already stopped! Cannot call stop again`);
+
+      stopped = true;
+      await root.shutdown();
+      await esServer.stop();
+    },
+
+    /**
+     * Migrates an array of SavedObjects and returns the results. Assumes that the objects will retain the same type
+     * and id after the migration. When testing migrations that may change a document's type or id, use `importObjects`
+     * and `exportObjects` directly.
+     * @param objects
+     */
+    migrate: async (objects: SavedObject[]) => {
+      await importObjects(objects);
+      return exportObjects({
+        objects: objects.map(({ type, id }) => ({ type, id })),
+      });
+    },
+
+    importObjects,
+    exportObjects,
+  };
+};
+
+export type SavedObjectTestHarness = ReturnType<typeof createTestHarness>;
+
+const waitForTrue = async (predicate: () => Promise<boolean>) => {
```

Comment:
optional: let's use named arguments from the very beginning. In the future, we might want to extend the function to make `maxAttempts` or `delay` configurable.